{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ['CUDA_VISIBLE_DEVICES'] = '1' # Must be before importing keras!\n",
    "import sys\n",
    "sys.path.insert(0,\"/home/kal/CTCF/modules/\")\n",
    "sys.path.insert(0,\"/home/kal/CTCF/mass_CTCF/modules/\")\n",
    "import tf_memory_limit\n",
    "#import matplotlib; matplotlib.use('Agg')  # Must be before importing matplotlib.pyplot or before importing pylab!\n",
    "from keras import backend as K\n",
    "import tensorflow as tf\n",
    "import numpy as np\n",
    "import ctcfgen\n",
    "import eval_convnet_ctcf\n",
    "import ucscgenome\n",
    "from convnet import Bias\n",
    "from keras.models import load_model, Model\n",
    "from keras.layers import Input, Activation\n",
    "import viz_sequence\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import pandas as pd\n",
    "from scipy.signal import correlate2d\n",
    "from sklearn.metrics import precision_recall_curve, roc_curve\n",
    "from scipy.special import logit\n",
    "\n",
    "colors = {0:'crimson', 1:'steelblue', 2:'goldenrod', 3:'forestgreen'}\n",
    "\n",
    "batch_size=32\n",
    "input_window=256\n",
    "\n",
    "meme_path = '/home/kal/data/CTCF.meme'\n",
    "gen_path='/home/kal/data/ctcfgen_data.hdf5'\n",
    "motif_bed_path = '/home/kal/data/1000genomes_motif_SNVs.bed'\n",
    "all_bed_path = '/home/kal/data/1000genomes_all_SNVs_motif.bed'\n",
    "\n",
    "# load the model\n",
    "model_path = '/home/kal/CTCF/mass_CTCF/output/sigmoid/20170919_150754/20170919_150754_weights_3_248_0.91.hdf5'\n",
    "model2_path = '/home/kal/CTCF/mass_CTCF/output/tanh/20170919_154126/20170919_154126_weights_3_241_0.90.hdf5'\n",
    "#model_path = '/home/kal/CTCF/mass_CTCF/eric_output/20170731_014309/20170731_014309_weights_3_243_0.91.hdf5'\n",
    "#model2_path = '/home/kal/CTCF/mass_CTCF/output/sigmoid/20170919_150754/20170919_150754_weights_3_248_0.91.hdf5'\n",
    "\n",
    "model = load_model(model_path, custom_objects={'Bias':Bias})\n",
    "use_act1 = True\n",
    "get_act = K.function([model.input, K.learning_phase()], [layer_dict['bias'].output])\n",
    "model2 = load_model(model2_path, custom_objects={'Bias':Bias})\n",
    "use_act2 = True\n",
    "\n",
    "# load the generator\n",
    "gen = ctcfgen.CTCFGeneratorhdf5(gen_path)\n",
    "\n",
    "#load the genome\n",
    "genome = ucscgenome.Genome('/home/kal/.ucscgenome/hg19.2bit')\n",
    "\n",
    "# get the symbolic outputs of each \"key\" layer.\n",
    "layer_dict = dict([(layer.name, layer) for layer in model.layers])\n",
    "layer_dict2 = dict([(layer.name, layer) for layer in model2.layers])\n",
    "print(layer_dict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def softmax(batch):\n",
    "    norm_batch = []\n",
    "    for seq in batch:\n",
    "        norm_batch.append([_softmax(base) for base in seq])\n",
    "    return np.asarray(norm_batch)\n",
    "    \n",
    "def _softmax(x):\n",
    "    e_x = np.exp(x - np.max(x))\n",
    "    return e_x / e_x.sum(axis=0) # only differenc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# process the 1000 genomes data\n",
    "snvs = pd.read_table(all_bed_path, header=None)\n",
    "snvs.columns = 'chr varstart varend TF ref A C G T motif_count'.split()\n",
    "snvs['motif'] = snvs['motif_count'] > 0\n",
    "snvs['start'] = snvs['varstart'] - 128\n",
    "snvs['end'] = snvs['varstart'] + 128\n",
    "order = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "rorder = ['A', 'C', 'G', 'T']\n",
    "ctcf_snvs = snvs[snvs.TF.apply(lambda x: 'CTCF' in x)].copy()\n",
    "ratios=list()\n",
    "alts=list()\n",
    "for idx, row in ctcf_snvs.iterrows():\n",
    "    counts = [row['A'], row['C'], row['G'], row['T']]\n",
    "    ref_count = counts[order[row['ref']]]\n",
    "    counts[order[row['ref']]] = 0\n",
    "    alt_count = max(counts)\n",
    "    alt = rorder[counts.index(alt_count)]\n",
    "    ratio = alt_count/(ref_count + alt_count)\n",
    "    alts.append(alt)\n",
    "    ratios.append(ratio)\n",
    "    \n",
    "ctcf_snvs['alt'] = np.asarray(alts)\n",
    "ctcf_snvs['allelicRatio'] = np.asarray(ratios)\n",
    "ctcf_snvs['mut'] = (ctcf_snvs.allelicRatio > .5)\n",
    "\n",
    "\n",
    "# load the bed\n",
    "#snvs = pd.read_table(motif_bed_path, header=None, sep=',')\n",
    "#snvs.columns = 'chr,varstart,varend,ref,alt,ref_freq,alt_freq,diff_freq,motif,TF,indiv_TF,allelicRatio'.split(',')\n",
    "#snvs['start'] = snvs['varstart'] - 128\n",
    "#snvs['end'] = snvs['varstart'] + 128\n",
    "\n",
    "#ctcf_snvs = snvs[snvs.TF.apply(lambda x: 'CTCF' in x)].copy()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# add annotations to the peaks\n",
    "for index, row in ctcf_snvs.iterrows():\n",
    "    # get the sequence\n",
    "    center = (row.start + row.end) // 2\n",
    "    if center < 128:\n",
    "        center = 128\n",
    "    orig_seq = ctcfgen.encode(np.fromstring(genome[row.chr][center-128:center+128].lower(), dtype=np.uint8))\n",
    "    new_seq = orig_seq.copy()\n",
    "    new_seq[128:129] = ctcfgen.encode_string(row['alt'].lower())\n",
    "    \n",
    "    # add the model scores\n",
    "    ctcf_snvs.set_value(index,'oml_score', get_act([np.asarray([orig_seq]*32),0])[0][0])\n",
    "    ctcf_snvs.set_value(index,'nml_score', get_act([np.asarray([new_seq]*32),0])[0][0])\n",
    "\n",
    "    # add the ctcf pwm scores\n",
    "    score, pwm, meme, meme_index, position = ctcfgen.get_pwm([orig_seq], get_everything=True)\n",
    "    ctcf_snvs.set_value(index,'octcf', score)\n",
    "    score, pwm, meme, meme_index, position = ctcfgen.get_pwm([new_seq], get_everything=True, meme=meme, position=position)\n",
    "    ctcf_snvs.set_value(index, 'nctcf', score)\n",
    "    ctcf_snvs.set_value(index,'ctcf_orentation', (meme_index>1))\n",
    "    ctcf_snvs.set_value(index,'ctcf_position', position)\n",
    "\n",
    "    \n",
    "    # add the mystery pwm scores\n",
    "    score, pwm, meme, meme_index, position = ctcfgen.get_pwm([orig_seq], get_everything=True, meme_library='mystery_memes')\n",
    "    ctcf_snvs.set_value(index,'omystery', score)\n",
    "    score, pwm, meme, meme_index, position = ctcfgen.get_pwm([new_seq], get_everything=True, meme=meme, position=position, meme_library='mystery_memes')\n",
    "    ctcf_snvs.set_value(index, 'nmystery', score)\n",
    "    ctcf_snvs.set_value(index,'mystery_orentation', (meme_index>0))\n",
    "    ctcf_snvs.set_value(index,'mystery_position', position)\n",
    "    \n",
    "ctcf_snvs['distance'] = abs(ctcf_snvs['mystery_position'] - ctcf_snvs['ctcf_position'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "orig_preds = list()\n",
    "new_preds = list()\n",
    "o_preds = list()\n",
    "n_preds = list()\n",
    "orig_pwms = list()\n",
    "new_pwms = list()\n",
    "in_pwm = list()\n",
    "mo_pwms=list()\n",
    "mn_pwms=list()\n",
    "\n",
    "if use_act1:\n",
    "    get_act = K.function([model.input, K.learning_phase()], [layer_dict['bias'].output])\n",
    "else:\n",
    "    get_act = K.function([model.input, K.learning_phase()], [model.output])\n",
    "    \n",
    "if use_act2:\n",
    "    get_act2 = K.function([model2.input, K.learning_phase()], [layer_dict2['bias'].output])\n",
    "else:\n",
    "    get_act2 = K.function([model2.input, K.learning_phase()], [model2.output])\n",
    "\n",
    "for index, row in ctcf_snvs.iterrows():\n",
    "    # get mutant version of the sequence\n",
    "    orig_seq = ctcfgen.encode(np.fromstring(genome[row.chr][row.start:row.end].lower(), dtype=np.uint8))\n",
    "    new_seq = orig_seq.copy()\n",
    "    new_seq[128:129] = ctcfgen.encode_string(row['alt'].lower())\n",
    "         \n",
    "    # get predictions\n",
    "    orig_pred = get_act([np.asarray([orig_seq]*32),0])[0][0]\n",
    "    new_pred = get_act([np.asarray([new_seq]*32),0])[0][0]\n",
    "    \n",
    "    #o_pred = get_act2([np.asarray([np.append(orig_seq, np.asarray([10]*256)).reshape(-1, 5)]*32),0])[0][0]\n",
    "    #n_pred = get_act2([np.asarray([np.append(new_seq, np.asarray([10]*256)).reshape(-1, 5)]*32),0])[0][0]\n",
    "    o_pred = get_act2([np.asarray([orig_seq]*32),0])[0][0]\n",
    "    n_pred = get_act2([np.asarray([new_seq]*32),0])[0][0]\n",
    "    \n",
    "    # get scores    \n",
    "    fseq, orig_score = ctcfgen.get_pwm([orig_seq], get_score=True)\n",
    "    fseq, new_score = ctcfgen.get_pwm([new_seq], get_score=True)   \n",
    "    \n",
    "    fseq, morig_score = ctcfgen.get_pwm([orig_seq], get_score=True, meme_library='mystery_memes')\n",
    "    fseq, mnew_score = ctcfgen.get_pwm([new_seq], get_score=True, meme_library='mystery_memes')  \n",
    "     \n",
    "    in_pwm.append(np.amax(fseq[118:138])>0)\n",
    "    \n",
    "    orig_preds.append(orig_pred)\n",
    "    new_preds.append(new_pred)\n",
    "    orig_pwms.append(orig_score)    \n",
    "    new_pwms.append(new_score)\n",
    "    mo_pwms.append(morig_score)\n",
    "    mn_pwms.append(mnew_score)\n",
    "    o_preds.append(o_pred)\n",
    "    n_preds.append(n_pred)\n",
    "    \n",
    "ctcf_snvs['refpred'] = np.asarray(orig_preds)\n",
    "ctcf_snvs['altpred'] = np.asarray(new_preds)\n",
    "\n",
    "ctcf_snvs['odiffpred'] = np.asarray(n_preds) - np.asarray(o_preds)\n",
    "\n",
    "ctcf_snvs['refscore'] = np.asarray(orig_pwms)\n",
    "ctcf_snvs['altscore'] = np.asarray(new_pwms)\n",
    "\n",
    "ctcf_snvs['mrefscore'] = np.asarray(mo_pwms)\n",
    "ctcf_snvs['maltscore'] = np.asarray(mn_pwms)\n",
    "\n",
    "ctcf_snvs['diffpred'] = ctcf_snvs['altpred']-ctcf_snvs['refpred']\n",
    "ctcf_snvs['diffscore'] = ctcf_snvs['altscore']-ctcf_snvs['refscore']\n",
    "ctcf_snvs['diffmscore'] = ctcf_snvs['maltscore']-ctcf_snvs['mrefscore']\n",
    "\n",
    "ctcf_snvs['pwm_motif'] = np.asarray(in_pwm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "filtered = ctcf_snvs\n",
    "#filtered = ctcf_snvs[~(ctcf_snvs['pwm_motif'])]\n",
    "#filtered = filtered[(filtered['diffpred']!=0)]\n",
    "ratios = filtered['allelicRatio'].tolist()\n",
    "pwms = filtered['diffscore'].tolist()\n",
    "mls = filtered['diffpred'].tolist()\n",
    "omls = filtered['odiffpred'].tolist()\n",
    "\n",
    "\n",
    "plt.hexbin(ratios, mls, gridsize=50, bins='log', cmap='plasma')\n",
    "plt.title('Single Nucleotide Varients in CTCF')\n",
    "plt.xlabel('Ratio Alternate/Total Reads')\n",
    "plt.ylabel('Difference in Sigmoid Model Score')\n",
    "plt.show()\n",
    "\n",
    "plt.hexbin(ratios, pwms, gridsize=50, bins='log', cmap='plasma')\n",
    "plt.title('Single Nucleotide Varients in CTCF')\n",
    "plt.xlabel('Ratio Alternate/Total Reads')\n",
    "plt.ylabel('Difference in PWM Score')\n",
    "plt.show()\n",
    "\n",
    "plt.hexbin(ratios, omls, gridsize=50, bins='log', cmap='plasma')\n",
    "plt.title('Single Nucleotide Varients in CTCF')\n",
    "plt.xlabel('Ratio Alternate/Total Reads')\n",
    "plt.ylabel('Difference in Tanh Model Score')\n",
    "plt.show()\n",
    "\n",
    "plt.hexbin(omls, mls, gridsize=50, bins='log', cmap='plasma')\n",
    "plt.title('Single Nucleotide Varients in CTCF')\n",
    "plt.xlabel('Difference in Tanh model Score')\n",
    "plt.ylabel('Difference in Sigmoid model Score')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# make histograms\n",
    "\n",
    "plt.hist(filtered[(filtered['mut'])]['diffscore'], 30, alpha=.5, label='Positives')\n",
    "plt.hist(filtered[~(filtered['mut'])]['diffscore'], 30, alpha=.5, label='Negatives')\n",
    "plt.title('Position Weight Matrix Prediction for CTCF binding')\n",
    "plt.xlabel('Position Weight Matrix Correlation')\n",
    "plt.legend()\n",
    "plt.show()\n",
    "\n",
    "\n",
    "plt.hist(filtered[(filtered['mut'])]['diffpred'], 30, alpha=.5, label='Positives')\n",
    "plt.hist(filtered[~(filtered['mut'])]['diffpred'], 30, alpha=.5, label='Negatives')\n",
    "plt.title('Deep Learning Sigmoid Model for CTCF binding')\n",
    "plt.xlabel('Deep Learning Score')\n",
    "plt.legend()\n",
    "plt.show()\n",
    "\n",
    "plt.hist(filtered[(filtered['mut'])]['odiffpred'], 30, alpha=.5, label='Positives')\n",
    "plt.hist(filtered[~(filtered['mut'])]['odiffpred'], 30, alpha=.5, label='Negatives')\n",
    "plt.title('Deep Learning Tanh Model for CTCF binding')\n",
    "plt.xlabel('Deep Learning Score')\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "deep_fpr, deep_tpr, thresholds = roc_curve(filtered['mut'].tolist(),\n",
    "                                                                 mls, pos_label=1)\n",
    "\n",
    "pwm_fpr, pwm_tpr, thresholds = roc_curve(filtered['mut'].tolist(),\n",
    "                                                                 pwms, pos_label=1)\n",
    "\n",
    "str_fpr, str_tpr, thresholds = roc_curve(filtered['mut'].tolist(),\n",
    "                                                                 omls, pos_label=1)\n",
    "\n",
    "plt.plot(deep_fpr, deep_tpr, label='Deep Learning')\n",
    "plt.plot(pwm_fpr, pwm_tpr, label = 'Position-Weight Matrix')\n",
    "plt.plot(str_fpr, str_tpr, label = 'Alternate Model')\n",
    "plt.legend()\n",
    "plt.title('ROC Curve')\n",
    "plt.xlabel('Recall')\n",
    "plt.ylabel('Precision')\n",
    "plt.ylim([0.0, 1.05])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.show()\n",
    "\n",
    "deep_precision, deep_recall, thresholds = precision_recall_curve(filtered['mut'].tolist(),\n",
    "                                                                 mls, pos_label=1)\n",
    "\n",
    "pwm_precision, pwm_recall, thresholds = precision_recall_curve(filtered['mut'].tolist(),\n",
    "                                                                 pwms, pos_label=1)\n",
    "\n",
    "str_precision, str_recall, thresholds = precision_recall_curve(filtered['mut'].tolist(),\n",
    "                                                                 omls, pos_label=1)\n",
    "\n",
    "plt.plot(deep_recall, deep_precision, label='Deep Learning')\n",
    "plt.plot(pwm_recall, pwm_precision, label = 'Position-Weight Matrix')\n",
    "plt.plot(str_recall, str_precision, label='Alternate Model')\n",
    "\n",
    "plt.legend()\n",
    "plt.title('Precision-Recall Curve')\n",
    "plt.xlabel('Recall')\n",
    "plt.ylabel('Precision')\n",
    "plt.ylim([0.0, 1.05])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.show()\n",
    "\n",
    "prop_pos = len(ctcf_snvs.loc[ctcf_snvs['mut'] == 1])/len(ctcf_snvs)\n",
    "print('Proportion positive:' + str(prop_pos))\n",
    "deep_precision_gain = [(x-prop_pos)/((1-prop_pos)*x) for x in deep_precision]\n",
    "deep_recall_gain = [(x-prop_pos)/((1-prop_pos)*x) for x in deep_recall]\n",
    "\n",
    "pwm_precision_gain = [(x-prop_pos)/((1-prop_pos)*x) for x in pwm_precision]\n",
    "pwm_recall_gain = [(x-prop_pos)/((1-prop_pos)*x) for x in pwm_recall]\n",
    "\n",
    "str_precision_gain = [(x-prop_pos)/((1-prop_pos)*x) for x in str_precision]\n",
    "str_recall_gain = [(x-prop_pos)/((1-prop_pos)*x) for x in str_recall]\n",
    "\n",
    "plt.plot(deep_recall_gain, deep_precision_gain, label='Deep Learning')\n",
    "plt.plot(pwm_recall_gain, pwm_precision_gain, label = 'Position-Weight Matrix')\n",
    "plt.plot(str_recall_gain, str_precision_gain, label = 'Alternate Model')\n",
    "\n",
    "plt.legend()\n",
    "plt.title('Precision-Recall Gain Curve')\n",
    "plt.xlabel('Recall')\n",
    "plt.ylabel('Precision')\n",
    "plt.ylim([0.0, 1.05])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def act_mutagenisis(seq, model=model):\n",
    "    get_act = K.function([model.input, K.learning_phase()], [layer_dict['bias'].output])\n",
    "\n",
    "    # do a mutant scan\n",
    "    mutant_window=1\n",
    "\n",
    "    # get a mutant batch generator\n",
    "    mutant_gen = ctcfgen.ngram_mutant_gen(seq, n=mutant_window)\n",
    "    g = ctcfgen.filled_batch(mutant_gen, batch_size=batch_size)\n",
    "    \n",
    "    # base importances as large-step gradients\n",
    "    # score with base there - average of scores without base there\n",
    "    mutant_preds = list()\n",
    "    for batch in g:\n",
    "        mutant_preds.append(get_act([batch, 0])[0])\n",
    "\n",
    "    return np.asarray(mutant_preds).reshape((-1, 4))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def get_importance(seq, my_model=model, use_act=True, start=0, end=256):\n",
    "    my_layer_dict = dict([(layer.name, layer) for layer in my_model.layers])\n",
    "    get_act = K.function([my_model.input, K.learning_phase()], [my_layer_dict['bias'].output])\n",
    "\n",
    "    pred = my_model.predict(np.asarray([seq]*32))[0][0]\n",
    "    if use_act:\n",
    "        score = get_act([[seq]*32, 0])[0][0][0]\n",
    "    else:\n",
    "        score = pred\n",
    "    print('Prediction ' + str(pred))\n",
    "\n",
    "    if use_act:\n",
    "        mutant_preds = act_mutagenisis(seq, model=model)\n",
    "    else:\n",
    "        mutant_preds = mutagenisis(seq, model=model)\n",
    "      \n",
    "    diffs = mutant_preds - score\n",
    "   \n",
    "    all_diffs = list()\n",
    "    for base_seq, base_preds in zip(seq, mutant_preds):\n",
    "        this_base = list()\n",
    "        for idx in range(4):\n",
    "            this_base.append(base_preds[idx] - np.average(base_preds))\n",
    "        all_diffs.append(list(this_base))\n",
    "\n",
    "    all_diffs=np.asarray(all_diffs)\n",
    "\n",
    "    score_diff = list()\n",
    "    for base_seq, base_preds in zip(seq, mutant_preds):\n",
    "        idx = np.where(base_seq)\n",
    "        score_diff.append(base_preds[idx] - np.average(base_preds))\n",
    "      \n",
    "    score_diff = (seq * score_diff)\n",
    "    \n",
    "    print('prediciton difference')\n",
    "    viz_sequence.plot_weights(all_diffs[start:end], colors=colors)\n",
    "\n",
    "    print('Masked average prediciton difference')\n",
    "    viz_sequence.plot_weights(score_diff[start:end], colors=colors)\n",
    "\n",
    "    #print('Softmax average prediction difference')\n",
    "    #viz_sequence.plot_icweights(softmax([all_diffs[start:end]])[0], colors=colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# find seqs of interest\n",
    "for index, row in ctcf_snvs.iterrows():\n",
    "    if (row.diffscore == 0) and ((abs(row.diffpred) > 1) or (abs(row.odiffpred) > 1)) and (((row.odiffpred>0) == (row.allelicRatio>.5)) or ((row.diffpred>0) == (row.allelicRatio>.5))) :\n",
    "    #if abs(row.odiffpred) > 200 and abs(row.diffpred) < .01 and ((row.odiffpred>0) == (row.allelicRatio>.5)) :\n",
    "        start=0\n",
    "        end=256\n",
    "        print(index)\n",
    "        \n",
    "        orig_seq = ctcfgen.encode(np.fromstring(genome[row.chr][row.start:row.end].lower(), dtype=np.uint8))\n",
    "        new_seq = orig_seq.copy()\n",
    "        new_seq[128:129] = ctcfgen.encode_string(row['alt'].lower())\n",
    "        \n",
    "        print('Sequence')\n",
    "        viz_sequence.plot_icweights(orig_seq[start:end], colors=colors)\n",
    "        print('Pred Difference ' + str(row.diffpred))\n",
    "        print('Strength Pred Difference: ' + str(row.odiffpred))\n",
    "        print('Score Difference ' + str(row.diffscore))\n",
    "        print('RC Difference ' + str(row.allelicRatio))\n",
    "        \n",
    "        print('Mutation')\n",
    "        viz_sequence.plot_weights((np.subtract(orig_seq,new_seq)*new_seq)[start:end], colors=colors)\n",
    "        fseq, orig_score = ctcfgen.get_pwm([orig_seq], get_score=True)\n",
    "        print('Original sequence weighting:')\n",
    "        get_importance(orig_seq, start=start, end=end)\n",
    "        get_importance(orig_seq, my_model=model2, start=start, end=end)\n",
    "        print('New Sequence weighting:')\n",
    "        get_importance(new_seq, start=start, end=end)\n",
    "        get_importance(new_seq, my_model=model2, start=start, end=end)\n",
    "        print('PWM ' + str(orig_score))\n",
    "        viz_sequence.plot_weights(fseq, colors=colors)\n",
    "        viz_sequence.plot_weights(fseq[start:end], colors=colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#new data set!\n",
    "bed_two = '/home/kal/data/IBD.CTCFpeak.bed'\n",
    "ibd = pd.read_table(bed_two)\n",
    "ibd.columns = 'HD chr region..mb. signal credible_nSNP tier2 trait trait.1 trait.reassigned sharingBF variant OR_CD OR_UC p_multi position A0 A1 AF_1KG info type certainty evidence.var evidence.set status P_mean_95 Func.refGene Gene.refGene GeneDetail.refGene ExonicFunc.refGene AAChange.refGene Coding TFBS epigenetic eQTL Disease CredibleSet CodingHuang SpliceSite Promoter pvalue TagSNP BestSNP PosteriorProb SeqA0 SeqA1'.split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "for index, row in ibd.iterrows():\n",
    "    if index == 30:\n",
    "        print(str(row))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "get_act = K.function([model.input, K.learning_phase()], [layer_dict['bias'].output])\n",
    "\n",
    "\n",
    "orig_preds = list()\n",
    "new_preds = list()\n",
    "o_preds = list()\n",
    "n_preds = list()\n",
    "orig_pwms = list()\n",
    "new_pwms = list()\n",
    "in_pwm = list()\n",
    "\n",
    "for index, row in ibd.iterrows():\n",
    "    \n",
    "    orig_seq = ctcfgen.encode_string(row['SeqA0'].lower())\n",
    "    orig_seq = orig_seq[:256]\n",
    "    new_seq = ctcfgen.encode_string(row['SeqA1'].lower())\n",
    "    new_seq = new_seq[:256]\n",
    "    \n",
    "    orig_pred = get_act([np.asarray([orig_seq]*32), 0])[0][0][0]\n",
    "    new_pred = get_act([np.asarray([new_seq]*32), 0])[0][0][0]\n",
    "    orig_preds.append(orig_pred)\n",
    "    new_preds.append(new_pred)\n",
    "    \n",
    "    o_preds.append(model2.predict(np.asarray([orig_seq]*32))[0][0])\n",
    "    n_preds.append(model2.predict(np.asarray([new_seq]*32))[0][0])\n",
    "    \n",
    "    fseq, orig_score = ctcfgen.get_pwm([orig_seq], get_score=True)\n",
    "    fseq, new_score = ctcfgen.get_pwm([new_seq], get_score=True)   \n",
    "    orig_pwms.append(orig_score)    \n",
    "    new_pwms.append(new_score)\n",
    "    \n",
    "    in_pwm.append(np.amax((np.subtract(orig_seq,new_seq)*new_seq)*fseq)>0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "ibd['refpred'] = np.asarray(orig_preds)\n",
    "ibd['altpred'] = np.asarray(new_preds)\n",
    "\n",
    "ibd['refscore'] = np.asarray(orig_pwms)\n",
    "ibd['altscore'] = np.asarray(new_pwms)\n",
    "\n",
    "ibd['diffpred'] = ibd['altpred']-ibd['refpred']\n",
    "ibd['diffscore'] = ibd['altscore']-ibd['refscore']\n",
    "\n",
    "ibd['odiffpred'] = np.asarray(n_preds) - np.asarray(o_preds)\n",
    "\n",
    "ibd['pwm_motif'] = np.asarray(in_pwm)\n",
    "\n",
    "ibd.to_csv('/home/kal/data/IBD.CTCFpeaks.ml.bed', sep='\\t', index=False, header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "def mutagenisis(seq, model=model):\n",
    "    # do a mutant scan\n",
    "    mutant_window=1\n",
    "\n",
    "    # get a mutant batch generator\n",
    "    mutant_gen = ctcfgen.ngram_mutant_gen(seq, n=mutant_window)\n",
    "    g = ctcfgen.filled_batch(mutant_gen, batch_size=batch_size)\n",
    "    \n",
    "    # base importances as large-step gradients\n",
    "    # score with base there - average of scores without base there\n",
    "    mutant_preds = list()\n",
    "    for batch in g:\n",
    "        mutant_preds.append(model.predict_on_batch(batch))\n",
    "\n",
    "    return np.asarray(mutant_preds).reshape((-1, 4))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "for index, row in ibd.iterrows():\n",
    "    if row.odiffpred != 0 or row.diffpred != 0:\n",
    "        start=0\n",
    "        end=256\n",
    "        print(index)\n",
    "        \n",
    "        orig_seq = ctcfgen.encode_string(row['SeqA0'].lower())\n",
    "        orig_seq = orig_seq[:256]\n",
    "        new_seq = ctcfgen.encode_string(row['SeqA1'].lower())\n",
    "        new_seq = new_seq[:256]\n",
    "        \n",
    "        print('Sequence')\n",
    "        viz_sequence.plot_icweights(orig_seq[start:end], colors=colors)\n",
    "        print('Sigmoid Difference ' + str(row.diffpred))\n",
    "        print('Tanh difference ' + str(row.odiffpred))\n",
    "        print('Score Difference ' + str(row.diffscore))\n",
    "        \n",
    "        print('Mutation')\n",
    "        viz_sequence.plot_weights((np.subtract(orig_seq,new_seq)*new_seq)[start:end], colors=colors)\n",
    "        fseq, orig_score = ctcfgen.get_pwm([orig_seq], get_score=True)\n",
    "        print('Sigmoid')\n",
    "        get_importance(orig_seq, start=start, end=end)\n",
    "        get_importance(new_seq, start=start, end=end)\n",
    "        print('Tanh')\n",
    "        get_importance(orig_seq, model=model2, start=start, end=end)\n",
    "        get_importance(new_seq, model=model2, start=start, end=end)\n",
    "        print('PWM ' + str(orig_score))\n",
    "        viz_sequence.plot_weights(fseq, colors=colors)\n",
    "        viz_sequence.plot_weights(fseq[start:end], colors=colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# process the memes\n",
    "def process_meme(meme_path):\n",
    "    with open(meme_path, 'r') as infile:\n",
    "        meme_length = -1\n",
    "        memes = list()\n",
    "        for line in infile.readlines():\n",
    "            if 'letter-probability matrix' in line:\n",
    "                meme_length = int(line.split()[5])\n",
    "                this_meme_lines = list()\n",
    "            elif meme_length > 0:\n",
    "                this_meme_lines.append([float(item.strip()) for item in line.split()])\n",
    "                meme_length = meme_length - 1\n",
    "            elif meme_length == 0:\n",
    "                this_meme = np.asarray(this_meme_lines)\n",
    "                memes.append(this_meme)\n",
    "                meme_length = -1\n",
    "        if meme_length == 0:\n",
    "            this_meme = np.asarray(this_meme_lines)\n",
    "            memes.append(this_meme)\n",
    "            meme_length = -1\n",
    "        \n",
    "    # add rcs of memes\n",
    "    rcs = list()\n",
    "    for meme in memes:\n",
    "        rcs.append(meme[::-1, ::-1])\n",
    "    memes = memes + rcs\n",
    "\n",
    "    \n",
    "    psuedocount=0.25\n",
    "    # get the transformed memes\n",
    "    transformed_memes = list()\n",
    "    for meme in memes:\n",
    "        meme = meme + psuedocount \n",
    "        norms = np.repeat(np.linalg.norm(meme, axis=1), 4).reshape((-1, 4))\n",
    "        meme = np.log(meme/norms)\n",
    "        min = np.amin(meme)\n",
    "        meme = meme - min\n",
    "        transformed_memes.append(meme)\n",
    "    \n",
    "    return transformed_memes\n",
    "\n",
    "CTCF_memes = process_meme('/home/kal/data/CTCF.meme')\n",
    "mystery_memes = process_meme('/home/kal/CTCF/mystery_seq/data/mystery_motif/mystery_motif.meme')\n",
    "\n",
    "viz_sequence.plot_weights(mystery_memes[0])\n",
    "\n",
    "def get_pwm(input_seqs, meme=None, position=None, meme_library=CTCF_memes, get_score=True, get_loc=False):  \n",
    "    # get position and meme if not specified\n",
    "    if meme == None and position == None:\n",
    "        get_loc = True\n",
    "        for seq in input_seqs:\n",
    "            best_score = -np.inf\n",
    "            for test_meme in meme_library:\n",
    "                correlations = correlate2d(seq, test_meme, mode='valid')\n",
    "                if np.max(correlations) > best_score:\n",
    "                    best_score = np.max(correlations)\n",
    "                    best_location = np.argmax(correlations)\n",
    "                    best_filter = test_meme\n",
    "        position = best_location\n",
    "        meme = best_filter\n",
    "    else:\n",
    "        get_loc = False\n",
    "    \n",
    "    # get the pwms\n",
    "    output_scores = list()\n",
    "    pwm = np.zeros(input_seqs[0].shape)\n",
    "    pwm[position:position+meme.shape[0]] = meme\n",
    "    for seq in input_seqs:\n",
    "        output_scores.append(np.copy(pwm))  \n",
    "    \n",
    "    if get_score:\n",
    "        corr = correlate2d(seq[position:position+meme.shape[0]], meme, mode='valid')\n",
    "        return pwm, np.max(corr)\n",
    "            \n",
    "    if get_loc:\n",
    "        return np.asarray(output_scores), meme, position\n",
    "    else:\n",
    "        return np.asarray(output_scores)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
